flink-connector-kudu/src/main/java/org/apache/flink/streaming/connectors/kudu/KuduSink.java

```diff
@@ -79,9 +86,12 @@ public KuduSink(String kuduMasters, KuduTableInfo tableInfo, KuduSerialization<O
 
     @Override
     public void open(Configuration parameters) throws IOException {
-        if (connector != null) return;
-        connector = new KuduConnector(kuduMasters, tableInfo, consistency, writeMode);
-        serializer.withSchema(tableInfo.getSchema());
+        if (this.connector != null) return;
+        FlushMode flushMode = ((StreamingRuntimeContext) getRuntimeContext()).isCheckpointingEnabled() ?
```

Comment:
I dont know if AUTO_FLUSH_BACKGROUND show be used here..

From Kudu documentacion
Note: The AUTO_FLUSH_BACKGROUND mode may result in out-of-order writes to Kudu. This is because in this mode multiple write operations may be sent to the server in parallel. See KUDU-1767 for more information.

Reply:
Now, the logic is :
-  If flink checkpoint is disable,synchronously write data to kudu.
-  If flink checkpoint is enable, asynchronously write data to kudu by default.
-  user also can change to sync by explicitly calling```KuduSink.withSyncFlushMode()``` when initializing KuduSink.  

the out-of-order tip is writed in ```KuduSink.withSyncFlushMode()``` comments, if someone want to orderly write data to kudu when checkpoint is enable, he can explicitly calling```KuduSink.withSyncFlushMode()``` .

Reply:
I have doubts if we should not merge these methods
-> withEventualConsistency - withAsyncFlushMode
-> withStrongConsistency - withSyncFlushMode

Reply:
you mean to unite consistency's config with flushmode's config ? 
-  ```ASYNC_MODE ( EVENTUAL + AUTO_FLUSH_BACKGROUND )```
-  ```SYNC_MODE ( STRONG + AUTO_FLUSH_SYNC )```

Reply:
yes.
its the same no?
the eventual consistence use async session and callback for response
the strong consistence force a response.join() to obtain the response (converting the call to a synchronous call)

if i understood correctly the auto_flush its doing the same no?

Reply:
In my understanding they are not equivalent.
- ```EVENTUAL``` and ```STRONG ``` mostly means out-of-order or ordered send message to Kudu.
- ```AUTO_FLUSH_BACKGROUND``` and ```AUTO_FLUSH_SYNC``` mostly means batch or single send message to Kudu.

```response.join()``` is Synchronously waits until this Deferred is called back.  If this Deferred already completed, this method returns (or throws) immediately. Otherwise, the current thread will be blocked and will wait until the Deferred is called back.
the  ```response.join()```  is only useful when you want to keep strong consistence In operation interleaving , eg: A opeartion and B opeartion need to be executed in order, when we use ```session.apply(operation)```  each opeartion will use client to send Rpc Request . in this step, there are sync, but maybe B opeartion is execute fast than A in kudu server，and we first got B opeartion 's callback. So it don't ensure consistence, only if we use ```response.join()```  to keep 
 waiting for A opeartion's callback before execute B opeartion.
But, whether we use ```response.join()``` or not. In default FlushMode ```session.apply(operation)`` is execute operation one by one , not batch. if we want to use buffer to impove throughput ,we can use another FlushMode like ```AUTO_FLUSH_BACKGROUND``` , and using checkpoint to ensure at-least-once.

Reply:
so the autofush will only make sense in eventual consistence no?

Reply:
the eventual consistence only make sense in AUTO_FLUSH_SYNC FlushMode.

Reply:
and strong consistence is ok with any flush mode?

Reply:
- ```AUTO_FLUSH_BACKGROUND``` don't support  strong consistence.
- ```MANUAL_FLUSH``` support  strong consistence by wait for one flush() and join() before triggering another flush. but in this mode, flush data only depend on checkpoint trigger. is it suitable in flink sink? should we add another strategy(count、size) ? we shoud do more in another PR.

Reply:
yes, maybe we have to rethink this.
we are forcing to chose strong or eventual consistence, but seems that some flush modes are not ok with any of these options..

could your create a new issue so we can discuss the best option

Reply:
ok, i will create a another issue to discuss it atfer a few days later.
And thanks for your review again.